Sources/CacheAdvance/CacheReader.swift

```diff
@@ -52,6 +55,18 @@ final class CacheReader {
 
         switch try nextEncodedMessageSpan() {
         case let .span(messageLength):
+            // Check our assumptions before we try to read the message.
+            let endOfMessage = Bytes(startingOffset) + Bytes(MessageSpan.storageLength) + Bytes(messageLength)
+            let startingOffsetIsBeforeEndOfNewestMessageAndDoesNotExceedEndOfNewestMessage = startingOffset < offsetInFileAtEndOfNewestMessage && endOfMessage <= offsetInFileAtEndOfNewestMessage
+            let startingOffsetIsOnOrAfterAfterEndOfNewestMessageAndDoesNotExceedEndOfFile = startingOffset >= offsetInFileAtEndOfNewestMessage && endOfMessage <= maximumBytes
+            guard
+                startingOffsetIsBeforeEndOfNewestMessageAndDoesNotExceedEndOfNewestMessage
+                    || startingOffsetIsOnOrAfterAfterEndOfNewestMessageAndDoesNotExceedEndOfFile
+            else {
+                // The offsetInFileAtEndOfNewestMessage is incorrect. This likely occured due to a crash when writing our header file.
+                throw CacheAdvanceError.fileCorrupted
```

Comment:
What do you think about throwing two errors to ensure that we have a test to cover each boolean separately? As-written we can't be as confident that our tests are covering all edge cases.

Reply:
I played with this in a debugger – we're not actually catching any issues with `startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile` (slightly updated name given the commit I just pushed, but you get the idea). What we _are_ doing, however, is ensuring that we pass the `guard` when `offsetInFileAtEndOfNewestMessage < startingOffset`.

Why aren't we catching issues with `startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile`? Because it's not a unicorn case to hit due to a crash. The only way for the `endOfMessage <= maximumBytes` check in `startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile` to be `false` is for `header.maximumBytes` to be smaller than expected, and we check that the `header.maximumBytes` in `canWriteToFile`:
https://github.com/dfed/CacheAdvance/blob/2bfee6c0f668735368bf8337cd726928002b1194/Sources/CacheAdvance/CacheHeaderHandle.swift#L160-L174

Granted, we only check `canWriteToFile` when we're appending messages, but since `header.maximumBytes` is never written outside of the initial header write... this seems not worth writing a test against? If our initial header write fails we have other problems.

Reply:
If I'm understanding correctly the only way `startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile` could catch issues is in the event of bit flip from a cosmic ray, or similar on-disk data corruption. Does that match your understanding?